projects.js

```diff
@@ -0,0 +1,226 @@
+class RepositoryProjectsManager {
+  constructor({ owner, repository, octokit }) {
+    this.owner = owner;
+    this.repositoryName = repository;
+    this.octokit = octokit;
+    this.clientMutationId = `assign-to-repository-projects-${owner}-${repository}`;
+  }
+
+  async assign(pullRequestId, titles) {
+    await this.#init();
+    await this.#assignPRToProjects(pullRequestId, titles);
+    await this.#unassignPRFromProjects(pullRequestId, titles);
+    return this.#assignedProjects(pullRequestId);
+  }
+
+  async #init() {
+    // the GitHub Action's event may only contain the "old" GraphQL node id.
+    // this produces deprecation warnings. as a workaround, look up the "new" ID.
+    // https://github.blog/changelog/label/deprecation/
+    const { organization } = await this.octokit.graphql(
+      `query {
+         organization(login: "${this.owner}") {
+           id
+           name
+         }
+      }`,
+      {
+        headers: {
+          'X-Github-Next-Global-ID': '1',
+        },
+      },
+    );
+    this.organization = organization;
+
+    await this.#fetchRepositoryAndProjects();
+  }
+
+  async #assignPRToProjects(pullRequestId, titles) {
+    const projects = this.projects.filter((p) => titles.includes(p.title));
+
+    for await (const project of projects) {
+      // async, because more than 5 breaks API endpoint
+      const item = await this.#assignPRtoProject(pullRequestId, project);
+
+      // at creation, items can only be assigned to projecs but initially
+      // have Status value null.
+      // the `Todo` Status value is created by default for each new project and
+      // must be assiged. to do so, the option value needs to be looked up first.
+      // (project automations cannot be used for this task because there is no API
+      // for creating or updating ProjectV2 workflows.)
+      await this.#assignStatusTodo(project, item);
+    }
+  }
+
+  async #fetchRepositoryAndProjects() {
+    const response = await this.octokit.graphql.paginate(`
+      query paginate($cursor: String) {
+        repository(owner: "${this.owner}", name: "${this.repositoryName}") {
+          name
+          id
+          projectsV2(first: 10, after: $cursor) {
+            nodes {
+              id
+              title
+              number
+              fields(first: 5) {
```

Comment:
why 5 as page size?

Reply:
I'll keep just this nested one because of the [arcane GraphQL limitations](https://docs.github.com/en/graphql/overview/resource-limitations). We know there should be fewer than 5 select fields by default. Actually there are 3 values for status. Should GitHub add one or two more we would still be fine.